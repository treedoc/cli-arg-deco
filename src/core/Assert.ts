export default class Assert {
  static isTrue(condition: boolean, error: string|(()=>string)) { 
    if (!condition) throw new Error(typeof(error) === 'string' ? error : error()); 
  }
} 
