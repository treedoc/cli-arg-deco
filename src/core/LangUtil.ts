export type Constructor = (...args: any[]) => any;  // Add just to avoid JS Lint complain
export function doIfNotNull<T>(obj: T | undefined, action: (obj: T) => void) { if (obj !== undefined && obj !== null)  action(obj); }
export function doIf(condition: boolean, action: () => void) { if (condition) action(); }

// TS_SPECIFIC
export function EnumValues(enumType: any) {
  return Object.keys(enumType).filter(key => !isNaN(Number(enumType[key])));
}

export function EnumsValueOf(enumType: any, value: number|string) {
  return enumType[value];
}

