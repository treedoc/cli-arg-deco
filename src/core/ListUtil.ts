export function isIn<T>(match: T, ...values: T[]) {
  if (values != null)
    for (const l of values) {
      if (match == null) {
        if (l == null)
          return true;
        continue;
      }
      if (match === l)
        return true;
    }
  return false;
}