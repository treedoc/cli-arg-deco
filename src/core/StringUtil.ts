export function noNull(s? : string) { return !s ? '' : s; }