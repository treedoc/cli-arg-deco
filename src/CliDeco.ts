import "reflect-metadata"
import { Constructor } from "./core/LangUtil";
import Param from "./Param";

const MK_NAME = Symbol("Cli:NAME");
const MK_DESCRIPTION = Symbol("Cli:DESCRIPTION");
const MK_SUMMARY = Symbol("Cli:SUMMARY");
const MK_EXAMPLES = Symbol("Cli:EXAMPLES");
const MK_INDEX = Symbol("Cli:INDEX");
const MK_SHORT_NAME = Symbol("Cli:SHORT_NAME");
const MK_REQUIRED = Symbol("Cli:REQUIRED");
const MK_DECODER = Symbol("Cli:DECODER");

const MK_DESIGN_TYPE = "design:type";
const EMPTY_MAP = new Map();

// Map structure: constructor|prototype -> propertyKey:string -> MetaKey:symbol? -> value:any
const metaMap = new Map<Constructor|object, Map<string|undefined, Map<symbol|undefined, any>>>();

function computerIfAbsent<K, V>(map: Map<K, V>, key: K, func: (key: K) => V): V {
  let result = map.get(key);  
  if (result === undefined) {
    result = func(key);
    map.set(key, result);
  }
  return result;
}

function getOrInitMap<K1, K2, V>(map: Map<K1, Map<K2, V>>, key: K1): Map<K2, V> {
  return computerIfAbsent(map, key, (k) => new Map());
}


function setMeta(target: Constructor|object, propertyKey: string | undefined, metaKey: symbol, val: any) {
  getOrInitMap(getOrInitMap(metaMap, target), propertyKey).set(metaKey, val);
}

function getMeta(target: Constructor|object, propertyKey: string | undefined, metaKey: symbol, val: any) {
  return metaMap.get(target)?.get(propertyKey)?.get(metaKey);
}


/**
 * Unfortunately, typescript refection function is very weak. There's no way to iterate the fields in the class if some fields
 * are not initialized explicitly. We override this method to record all the field withe the Metadata decorations
 */
function metadata(metadataKey: any, val: any) {
  const result = Reflect.metadata(metadataKey, val);
  return (target: Constructor | object, propertyKey?: string) => {
    // For now, we store the mete information both locally and with Reflect.js
    setMeta(target, propertyKey, metadataKey, val);
    propertyKey === undefined ? result(target as Constructor) : result(target, propertyKey);
  }
}

/**
 * Reflect has inconsistent use of target for field and class decoration.
 * For class, it's using the constructor function
 * For field, it's using contractor.prototype. 
 * This method hide this inconsistency. 
 */
function getMetadata(metadataKey: any, target: Constructor, propertyKey?: string | symbol) {
  return propertyKey === undefined ? Reflect.getMetadata(metadataKey, target)
    : Reflect.getMetadata(metadataKey, target.prototype, propertyKey);
}

export default class CliDeco {
  static Name(val: string) { return metadata(MK_NAME, val); }
  static Description(val: string) { return metadata(MK_DESCRIPTION, val); }
  static Summary(val: string) { return metadata(MK_SUMMARY, val); }
  static Examples(val: string[]) { return metadata(MK_EXAMPLES, val); }
  static Index(val: number) { return metadata(MK_INDEX, val); }
  static ShortName(val: string) { return metadata(MK_SHORT_NAME, val); }
  static Required(val: boolean) { return metadata(MK_REQUIRED, val); }
  static Decoder(val: (s: string) => any) { return metadata(MK_DECODER, val); }

  static getName(target: Constructor, key?: string): string | undefined { return getMetadata(MK_NAME, target, key); }
  static getDescription(target: Constructor, key?: string): string | undefined { return getMetadata(MK_DESCRIPTION, target, key); }
  static getSummary(target: Constructor, key?: string): string | undefined { return getMetadata(MK_SUMMARY, target, key); }
  static getExamples(target: Constructor, key?: string): string[] | undefined { return getMetadata(MK_EXAMPLES, target, key); }
  static getIndex(target: Constructor, key?: string): number | undefined { return getMetadata(MK_INDEX, target, key); }
  static getShortName(target: Constructor, key?: string): string | undefined { return getMetadata(MK_SHORT_NAME, target,  key); }
  static getRequired(target: Constructor, key?: string): boolean | undefined { return getMetadata(MK_REQUIRED, target, key); }
  static getDecoder(target: Constructor, key?: string): (s: string) => any | undefined { return getMetadata(MK_DECODER, target, key); }

  static getDesignType(target: Constructor, key?: string): boolean | undefined { return getMetadata(MK_DESIGN_TYPE, target, key); }

  static getFieldsMetadata(target: Constructor): Map<string, Map<symbol|undefined, any>> {
    return metaMap.get(target.prototype) || EMPTY_MAP;
  }

  static getMetadataForField(target: Constructor, key: string | symbol, metaKey: symbol) {
    return getMetadata(metaKey, target,  key);
  }
}

