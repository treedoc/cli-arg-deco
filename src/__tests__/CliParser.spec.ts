import CliDeco from "../CliDeco"
import CLISpec from "../CliSpec";
import { EnumsValueOf, EnumValues } from "../core/LangUtil";


const { Name, Summary, Description, Examples, Index, ShortName, Required, Decoder } = CliDeco;

class Point {
  name?: string;
  x?: number;
  y?: number;
}

enum Opt{ VAL1, VAL2 }

@Name("TestArg1") 
@Summary("This is a test arg1") 
@Description("Description of test Args")
@Examples([
    "arg1 2",
    "arg1 4",
])
class Arg1 {
  @Index(0) @Description("Str parameter")
  strParam?: string;

  @Index(1) @Description("number parameter")
  numParam?: number;

  @Index(2) @Description("Object Point") @Required(false)
  point?: Point;

  // TODO: Add enum support, Typescript metadata information of `design:type` will set as `object` for enum which is not correct
  @ShortName("o") @Description( `Opt: possibleValues:[${EnumValues(Opt)}]`) @Required(true) @Decoder(_ => EnumsValueOf(Opt, _))
  opt = Opt.VAL1;

  @ShortName("i")
  optInt: number = 10;
}


describe('CliParser', () => {
  test('testParse', () => {
    const spec = new CLISpec(Arg1);
    expect(spec).toMatchSnapshot("spec");
    expect(spec.usage).toMatchSnapshot("usage");
    console.info("spec:\n" + spec.printUsage());
    expect(spec.printUsage()).toMatchSnapshot("printUsage");

    const args = ["abc", "10", 'name:n1,x:1,y:2', "-o", "VAL2", "--optInt", "100"];
    const parser = spec.parse(args, 0);
    console.info("parsedValue:\n" + JSON.stringify(parser.target, null, 2));
    expect(parser.target).toMatchSnapshot("parserTarget");
  });
});