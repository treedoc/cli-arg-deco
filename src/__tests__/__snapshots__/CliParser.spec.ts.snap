// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CliParser testParse: parserTarget 1`] = `
Arg1 {
  "numParam": 10,
  "opt": 1,
  "optInt": 100,
  "point": Object {
    "name": "n1",
    "x": 1,
    "y": 2,
  },
  "strParam": "abc",
}
`;

exports[`CliParser testParse: printUsage 1`] = `
"NAME: TestArg1 - This is a test arg1
DESCRIPTION
  Description of test Args
USAGE
  TestArg1 -o {opt} [-i {optInt}] <strParam> <numParam> [point]
EXAMPLES
  arg1 2
  arg1 4
ARGUMENTS / OPTIONS
  <strParam>:   Str parameter
  <numParam>:   number parameter
  <point>:      Object Point
  -o, --opt:    Opt: possibleValues:[VAL1,VAL2]
  -i, --optInt: 
"
`;

exports[`CliParser testParse: spec 1`] = `
CLISpec {
  "cls": [Function],
  "defVal": Arg1 {
    "opt": 0,
    "optInt": 10,
  },
  "description": "Description of test Args",
  "examples": Array [
    "arg1 2",
    "arg1 4",
  ],
  "firstOptionalIndex": 2,
  "indexedParams": Array [
    Param {
      "cls": [Function],
      "defVal": undefined,
      "description": "Str parameter",
      "index": 0,
      "key": "strParam",
      "name": "strParam",
      "type": [Function],
    },
    Param {
      "cls": [Function],
      "defVal": undefined,
      "description": "number parameter",
      "index": 1,
      "key": "numParam",
      "name": "numParam",
      "type": [Function],
    },
    Param {
      "cls": [Function],
      "defVal": undefined,
      "description": "Object Point",
      "index": 2,
      "key": "point",
      "name": "point",
      "required": false,
      "type": [Function],
    },
  ],
  "name": "TestArg1",
  "optionParams": Array [
    Param {
      "cls": [Function],
      "decoder": [Function],
      "defVal": 0,
      "description": "Opt: possibleValues:[VAL1,VAL2]",
      "key": "opt",
      "name": "opt",
      "required": true,
      "shortName": "o",
      "type": [Function],
    },
    Param {
      "cls": [Function],
      "defVal": 10,
      "key": "optInt",
      "name": "optInt",
      "shortName": "i",
      "type": [Function],
    },
  ],
  "requiredParams": Set {
    "strParam",
    "numParam",
    "opt",
  },
  "summary": "This is a test arg1",
}
`;

exports[`CliParser testParse: usage 1`] = `"TestArg1 -o {opt} [-i {optInt}] <strParam> <numParam> [point]"`;
